File: .skills/openclaw-skills/skills/obcraft/apiosk/apiosk_client.py
"""
Apiosk Python Client
Easy API calls with automatic x402 payment
"""

import os
import json
import requests
from pathlib import Path

WALLET_FILE = Path.home() / '.apiosk' / 'wallet.json'
CONFIG_FILE = Path.home() / '.apiosk' / 'config.json'


def load_wallet():
    """Load wallet configuration"""
    if not WALLET_FILE.exists():
        raise FileNotFoundError('Wallet not found. Run ./setup-wallet.sh first')
    
    with open(WALLET_FILE) as f:
        wallet = json.load(f)
    
    with open(CONFIG_FILE) as f:
        config = json.load(f)
    
    return wallet, config


def call_apiosk(api_id: str, params: dict = None) -> dict:
    """
    Call an Apiosk API
    
    Args:
        api_id: API identifier (e.g., 'weather', 'prices')
        params: API parameters
    
    Returns:
        API response as dictionary
    
    Raises:
        requests.HTTPError: If API call fails
        ValueError: If payment fails
    """
    wallet, config = load_wallet()
    params = params or {}
    
    url = f"{config['gateway_url']}/{api_id}"
    headers = {
        'Content-Type': 'application/json',
        'X-Wallet-Address': wallet['address'],
    }
    
    response = requests.post(url, json=params, headers=headers)
    
    if response.status_code == 200:
        return response.json()
    elif response.status_code == 402:
        error = response.json().get('error', 'Insufficient balance')
        raise ValueError(f'Payment required: {error}')
    else:
        response.raise_for_status()


def list_apis() -> list:
    """
    List available APIs
    
    Returns:
        List of API dictionaries
    """
    _, config = load_wallet()
    
    url = f"{config['gateway_url']}/v1/apis"
    response = requests.get(url)
    response.raise_for_status()
    
    return response.json()['apis']


def check_balance() -> dict:
    """
    Check wallet balance
    
    Returns:
        Balance information dictionary
    """
    wallet, config = load_wallet()
    
    url = f"{config['gateway_url']}/v1/balance"
    params = {'address': wallet['address']}
    
    response = requests.get(url, params=params)
    response.raise_for_status()
    
    return response.json()


def usage_stats(period: str = 'all') -> dict:
    """
    Get usage statistics
    
    Args:
        period: 'all', 'today', 'week', or 'month'
    
    Returns:
        Usage statistics dictionary
    """
    wallet, config = load_wallet()
    
    url = f"{config['gateway_url']}/v1/usage"
    params = {
        'address': wallet['address'],
        'period': period,
    }
    
    response = requests.get(url, params=params)
    response.raise_for_status()
    
    return response.json()


# Example usage
if __name__ == '__main__':
    print('🦞 Apiosk Client Example\n')
    
    try:
        # List APIs
        print('Available APIs:')
        apis = list_apis()
        for api in apis:
            print(f"- {api['id']}: ${api['price_usd']}/req - {api['description']}")
        
        print('\n')
        
        # Call weather API
        print('Calling weather API for Amsterdam...')
        weather = call_apiosk('weather', {'city': 'Amsterdam'})
        print(f"Temperature: {weather['temperature']}°C")
        print(f"Condition: {weather['condition']}")
        print('✅ Paid: $0.001 USDC\n')
        
        # Check balance
        balance = check_balance()
        print(f"Remaining balance: ${balance['balance_usdc']} USDC")
        
    except Exception as e:
        print(f'❌ Error: {e}')
        exit(1)

File: .skills/openclaw-skills/skills/obcraft/apiosk/apiosk-client.js
/**
 * Apiosk Node.js Client
 * Easy API calls with automatic x402 payment
 */

const fs = require('fs');
const path = require('path');
const https = require('https');

const WALLET_FILE = path.join(process.env.HOME, '.apiosk', 'wallet.json');
const CONFIG_FILE = path.join(process.env.HOME, '.apiosk', 'config.json');

/**
 * Load wallet configuration
 */
function loadWallet() {
  if (!fs.existsSync(WALLET_FILE)) {
    throw new Error('Wallet not found. Run ./setup-wallet.sh first');
  }
  
  const wallet = JSON.parse(fs.readFileSync(WALLET_FILE, 'utf8'));
  const config = JSON.parse(fs.readFileSync(CONFIG_FILE, 'utf8'));
  
  return { wallet, config };
}

/**
 * Call an Apiosk API
 * @param {string} apiId - API identifier (e.g., 'weather', 'prices')
 * @param {object} params - API parameters
 * @returns {Promise<object>} API response
 */
async function callApiosk(apiId, params = {}) {
  const { wallet, config } = loadWallet();
  
  const url = new URL(`/${apiId}`, config.gateway_url);
  
  return new Promise((resolve, reject) => {
    const data = JSON.stringify(params);
    
    const options = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': data.length,
        'X-Wallet-Address': wallet.address,
      }
    };
    
    const req = https.request(url, options, (res) => {
      let body = '';
      
      res.on('data', (chunk) => {
        body += chunk;
      });
      
      res.on('end', () => {
        try {
          const response = JSON.parse(body);
          
          if (res.statusCode === 200) {
            resolve(response);
          } else if (res.statusCode === 402) {
            reject(new Error(`Payment required: ${response.error || 'Insufficient balance'}`));
          } else {
            reject(new Error(`API error (${res.statusCode}): ${response.error || body}`));
          }
        } catch (e) {
          reject(new Error(`Failed to parse response: ${e.message}`));
        }
      });
    });
    
    req.on('error', (e) => {
      reject(new Error(`Request failed: ${e.message}`));
    });
    
    req.write(data);
    req.end();
  });
}

/**
 * List available APIs
 * @returns {Promise<Array>} List of APIs
 */
async function listAPIs() {
  const { config } = loadWallet();
  const url = new URL('/v1/apis', config.gateway_url);
  
  return new Promise((resolve, reject) => {
    https.get(url, (res) => {
      let body = '';
      
      res.on('data', (chunk) => {
        body += chunk;
      });
      
      res.on('end', () => {
        try {
          const data = JSON.parse(body);
          resolve(data.apis);
        } catch (e) {
          reject(new Error(`Failed to parse response: ${e.message}`));
        }
      });
    }).on('error', (e) => {
      reject(new Error(`Request failed: ${e.message}`));
    });
  });
}

/**
 * Check wallet balance
 * @returns {Promise<object>} Balance info
 */
async function checkBalance() {
  const { wallet, config } = loadWallet();
  const url = new URL(`/v1/balance?address=${wallet.address}`, config.gateway_url);
  
  return new Promise((resolve, reject) => {
    https.get(url, (res) => {
      let body = '';
      
      res.on('data', (chunk) => {
        body += chunk;
      });
      
      res.on('end', () => {
        try {
          const data = JSON.parse(body);
          resolve(data);
        } catch (e) {
          reject(new Error(`Failed to parse response: ${e.message}`));
        }
      });
    }).on('error', (e) => {
      reject(new Error(`Request failed: ${e.message}`));
    });
  });
}

module.exports = {
  callApiosk,
  listAPIs,
  checkBalance,
};

// Example usage
if (require.main === module) {
  (async () => {
    try {
      console.log('🦞 Apiosk Client Example\n');
      
      // List APIs
      console.log('Available APIs:');
      const apis = await listAPIs();
      apis.forEach(api => {
        console.log(`- ${api.id}: $${api.price_usd}/req - ${api.description}`);
      });
      
      console.log('\n');
      
      // Call weather API
      console.log('Calling weather API for Amsterdam...');
      const weather = await callApiosk('weather', { city: 'Amsterdam' });
      console.log(`Temperature: ${weather.temperature}°C`);
      console.log(`Condition: ${weather.condition}`);
      console.log('✅ Paid: $0.001 USDC\n');
      
      // Check balance
      const balance = await checkBalance();
      console.log(`Remaining balance: $${balance.balance_usdc} USDC`);
      
    } catch (error) {
      console.error('❌ Error:', error.message);
      process.exit(1);
    }
  })();
}
